### PR TITLE
feat(464): add support for queueing async actions in background

### PR DESCRIPTION
**Pull request by Hamsajj**
# Ticket(s)
#464 

## Description

This is an alternative approach to this https://github.com/gatewayd-io/gatewayd/pull/519 which does not rely on any third-party library.

Uses Redis Pub/Sub as a backend for queuing async actions (if configured).
It will fall back to the normal goroutine if not configured to use Redis. 

A consumer is added to [gateway-plugin-sdk](https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/34), which is used in tests to show how it can be used. 

## Related PRs
This PR is dependent to this one being merged and released https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/34

### CODE REVIEW

**Comment by Hamsajj** [comment]
This is in draft mode because it's tests need https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/34 to be merged and released

**Comment by mostafa** [comment]
@Hamsajj I update the SDK and dependencies and you need to rebase from `main`. I already did it locally, but I don't have permission to update your fork.

File: config/config.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ type IConfig interface {
 type Config struct {
 	globalDefaults GlobalConfig
 	pluginDefaults PluginConfig
+	actDefaults    ActConfig
```
[the remark this answers is not in the document]
Comment (Hamsajj) [reply config/config.go:39]: I thought that async actions (sending data to a background queue) are not directly related to plugins and are a gatewayd configuration. But tbh I did not really thought about it and I don't have any strong reasoning.

and looking more at the yaml file, it make sense to keep them there as well. will move the config back.

[the remark this answers is not in the document]
Comment (mostafa) [reply config/config.go]: We might want to consolidate the two config files into one at some point, but for now it's okay to keep this as you've changed it (to be in the `gatewayd_plugins.yaml` file).

File: act/registry.go [hunk 2]
```diff
@@ -26,6 +28,10 @@ type Registry struct {
 	// Default timeout for running actions
 	DefaultActionTimeout time.Duration
 
+	// TaskPublisher is the publisher for async actions.
+	// if not given, will invoke simple goroutine to run async actions
+	TaskPublisher *Publisher
```
Comment (mostafa) [inline act/registry.go:33]: You were using an interface here, which causes the pointer not to work as expected, and I fixed it by using to the concrete type.

File: config/config.go [hunk 3]
```diff
@@ -237,6 +237,11 @@ func (c *Config) LoadDefaults(ctx context.Context) *gerr.GatewayDError {
 		PolicyTimeout:       DefaultPolicyTimeout,
 		ActionTimeout:       DefaultActionTimeout,
 		Policies:            []Policy{},
+		ActionRedis: ActionRedisConfig{
+			Enabled: DefaultActionRedisEnabled,
+			Address: DefaultRedisAddress,
+			Channel: DefaultRedisChannel,
+		},
```
Comment (mostafa) [inline config/config.go:244]: Added default values here.

**Review by mostafa — APPROVED** [review]
@Hamsajj Thanks for your awesome contribution! :pray:

I tested it locally and found and fixed a bug and did some minor adjustments.

File: config/config.go (hunk 3 above)
Comment (Hamsajj) [reply config/config.go:244]: thanks. I thought the zero values are good enough default values 😄

**Pull request merged**